repo/ecc/ecc_rs_crc.go

```diff
@@ -0,0 +1,430 @@
+	dataBytes := inputBytes[parityPlusCrcSizeInBlock*sizes.Blocks:]
+
+	var maxShards [256][]byte
+	shards := maxShards[:sizes.DataShards+sizes.ParityShards]
+
+	dataPos := 0
+	eccPos := 0
+
+	var originalSize int
+
+	writeOriginalPos := 0
+	paddingStartPos := len(copied) - parityPlusCrcSizeInBlock*sizes.Blocks
+
+	for b := 0; b < sizes.Blocks; b++ {
+		for i := 0; i < sizes.DataShards; i++ {
+			initialDataPos := dataPos
+
+			crc := binary.BigEndian.Uint32(dataBytes[dataPos : dataPos+crcSize])
+			dataPos += crcSize
+
+			shards[i] = dataBytes[dataPos : dataPos+sizes.ShardSize]
+			dataPos += sizes.ShardSize
+
+			// We don't need to compute the crc inside the padding
+			if initialDataPos < paddingStartPos {
+				if crc != crc32.ChecksumIEEE(shards[i]) {
+					// The data was corrupted, so we need to reconstruct it
+					shards[i] = nil
+				}
+			}
+		}
+
+		for i := 0; i < sizes.ParityShards; i++ {
+			s := sizes.DataShards + i
+
+			crc := binary.BigEndian.Uint32(eccBytes[eccPos : eccPos+crcSize])
+			eccPos += crcSize
+
+			shards[s] = eccBytes[eccPos : eccPos+sizes.ShardSize]
+			eccPos += sizes.ShardSize
+
+			if crc != crc32.ChecksumIEEE(shards[s]) {
+				// The data was corrupted, so we need to reconstruct it
+				shards[s] = nil
+			}
+		}
+
+		if r.Options.DeleteFirstShardForTests {
+			shards[0] = nil
+		}
+
+		err := sizes.enc.ReconstructData(shards)
+		if err != nil {
+			return errors.Wrap(err, "Error computing ECC")
+		}
+
+		startShard := 0
+		startByte := 0
+
+		if b == 0 {
+			originalSize, startShard, startByte = readLength(shards, &sizes)
+		}
+
+		// Copy data to output. Because we prepend the original file length to the
+		// data, we need to ignore it
+		for i := startShard; i < sizes.DataShards && writeOriginalPos < originalSize; i++ {
+			left := minInt(originalSize-writeOriginalPos, sizes.ShardSize-startByte)
+			writeOriginalPos += left
+
+			output.Append(shards[i][startByte : startByte+left])
+
+			startByte = 0
+		}
+	}
+
+	return nil
+}
+
+func readLength(shards [][]byte, sizes *sizesInfo) (originalSize, startShard, startByte int) {
+	var lengthBuffer [lengthSize]byte
+
+	switch sizes.ShardSize {
+	case 1:
+		startShard = 4
+		startByte = 0
+
+		for i := 0; i < 4; i++ {
+			lengthBuffer[i] = shards[i][0]
+		}
+
+	case 2: //nolint:gomnd
+		startShard = 2
+		startByte = 0
+
+		copy(lengthBuffer[0:2], shards[0])
+		copy(lengthBuffer[2:4], shards[1])
+
+	case 3: //nolint:gomnd
+		startShard = 1
+		startByte = 1
+
+		copy(lengthBuffer[0:3], shards[0])
+		copy(lengthBuffer[3:4], shards[1])
+
+	case 4: //nolint:gomnd
+		startShard = 1
+		startByte = 0
+
+		copy(lengthBuffer[0:4], shards[0])
+
+	default:
+		startShard = 0
+		startByte = 4
+
+		copy(lengthBuffer[0:4], shards[0][:4])
+	}
+
+	originalSize = int(binary.BigEndian.Uint32(lengthBuffer[:]))
+
+	//nolint:nakedret
+	return
+}
+
+// Overhead is not used. It's just implemented to avoid errors when running.
+func (r *ReedSolomonCrcECC) Overhead() int {
+	return 0
```

Comment:
we should panic here, so we remember to never call this.

Reply:
Currently this method is called by `content.SharedManager.setupReadManagerCaches`:
```
	sm.committedContents = newCommittedContentIndex(caching,
		uint32(sm.format.Encryptor().Overhead()),
		sm.format.WriteIndexVersion(),
		sm.enc.getEncryptedBlob,
		sm.namedLogger("committed-content-index"),
		caching.MinIndexSweepAge.DurationOrDefault(DefaultIndexCacheSweepAge))
```

I can panic it now, because it is not being called, but when integrating this will be a problem.

Reply:
that's ok, when we integrate this we will need to fix this logic.